File: src/bind/registration.cpp
#include <pcl/registration/ia_ransac.h>
#include <pcl/registration/icp.h>
#include <emscripten/val.h>
#include <Eigen/Dense>

#include "embind.hpp"

using namespace pcl;
using namespace emscripten;
thread_local const val Float32Array = emscripten::val::global("Float32Array");


template <typename PointSource,typename PointTarget>
val getFinalTransformation(pcl::Registration<PointSource,PointTarget> &icp) {
    Eigen::Matrix4f transformation = icp.getFinalTransformation();
    auto js_result = val::null();
    std::array<float, 16> array;
    int idx = 0;
    for (int col = 0; col < transformation.cols(); ++col) {
        for (int row = 0; row < transformation.rows(); ++row) {
            array[idx++] = transformation(row, col);
        }
    }
    js_result = val(typed_memory_view(array.size(), array.data()));
    return js_result;
}

template <typename PointSource,typename PointTarget>
void align(pcl::IterativeClosestPoint<PointSource,PointTarget> &icp,PointCloud<PointTarget> &aligned,val array) {
    unsigned int length = array["length"].as<float>();
    if (length != 16) {
        icp.align(aligned);
        return;
    }else{
        Eigen::Matrix4f mat;
        for (int i = 0; i < 4; ++i) {
            for (int j = 0; j < 4; ++j) {
                mat(i, j) =  array[i * 4 + j].as<float>();
            }
        }
        icp.align(aligned,mat);
        return;
    }
}




#define BIND_Registration(PointSource, PointTarget)                                               \
  class_<pcl::Registration<PointSource, PointTarget>>("Registration" #PointSource #PointTarget)   \
      .function("hasConverged", &pcl::Registration<PointSource, PointTarget>::hasConverged)       \
      .function("getInputSource", &pcl::Registration<PointSource, PointTarget>::getInputSource)   \
      .function("getInputTarget", &pcl::Registration<PointSource, PointTarget>::getInputTarget)   \
      .function("getFitnessScore",                                                                \
                select_overload<double(double)>(                                                  \
                    &pcl::Registration<PointSource, PointTarget>::getFitnessScore))               \
      .function("setMaximumIterations",                                                           \
                &pcl::Registration<PointSource, PointTarget>::setMaximumIterations)               \
      .function("getMaximumIterations",                                                           \
                &pcl::Registration<PointSource, PointTarget>::getMaximumIterations)               \
      .function("setRANSACIterations",                                                            \
                &pcl::Registration<PointSource, PointTarget>::setRANSACIterations)                \
      .function("getRANSACIterations",                                                            \
                &pcl::Registration<PointSource, PointTarget>::getRANSACIterations)                \
      .function("setRANSACOutlierRejectionThreshold",                                             \
                &pcl::Registration<PointSource, PointTarget>::setRANSACOutlierRejectionThreshold) \
      .function("getRANSACOutlierRejectionThreshold",                                             \
                &pcl::Registration<PointSource, PointTarget>::getRANSACOutlierRejectionThreshold) \
      .function("setMaxCorrespondenceDistance",                                                   \
                &pcl::Registration<PointSource, PointTarget>::setMaxCorrespondenceDistance)       \
      .function("getMaxCorrespondenceDistance",                                                   \
                &pcl::Registration<PointSource, PointTarget>::getMaxCorrespondenceDistance)       \
      .function("setTransformationEpsilon",                                                       \
                &pcl::Registration<PointSource, PointTarget>::setTransformationEpsilon)           \
      .function("getTransformationEpsilon",                                                       \
                &pcl::Registration<PointSource, PointTarget>::getTransformationEpsilon)           \
      .function("setTransformationRotationEpsilon",                                               \
                &pcl::Registration<PointSource, PointTarget>::setTransformationRotationEpsilon)   \
      .function("getTransformationRotationEpsilon",                                               \
                &pcl::Registration<PointSource, PointTarget>::getTransformationRotationEpsilon)   \
      .function("setEuclideanFitnessEpsilon",                                                     \
                &pcl::Registration<PointSource, PointTarget>::setEuclideanFitnessEpsilon)         \
      .function("getEuclideanFitnessEpsilon",                                                     \
                &pcl::Registration<PointSource, PointTarget>::getEuclideanFitnessEpsilon)         \
      .function("initCompute", &pcl::Registration<PointSource, PointTarget>::initCompute)         \
      .function("initComputeReciprocal",                                                          \
                &pcl::Registration<PointSource, PointTarget>::initComputeReciprocal)              \
      .function(                                                                                  \
          "align",                                                                                \
          &align<PointSource, PointTarget>)                                                       \
      .function("getFinalTransformation",&getFinalTransformation<PointSource, PointTarget>);

#define BIND_IterativeClosestPoint(PointSource, PointTarget)                                      \
  class_<pcl::IterativeClosestPoint<PointSource, PointTarget>,                                    \
         base<pcl::Registration<PointSource, PointTarget>>>(                                      \
      "IterativeClosestPoint" #PointSource #PointTarget)                                          \
      .constructor()                                                                              \
      .function("setInputSource",                                                                 \
                &pcl::IterativeClosestPoint<PointSource, PointTarget>::setInputSource)            \
      .function("setInputTarget",                                                                 \
                &pcl::IterativeClosestPoint<PointSource, PointTarget>::setInputTarget)            \
      .function(                                                                                  \
          "setUseReciprocalCorrespondences",                                                      \
          &pcl::IterativeClosestPoint<PointSource, PointTarget>::setUseReciprocalCorrespondences) \
      .function(                                                                                  \
          "getUseReciprocalCorrespondences",                                                      \
          &pcl::IterativeClosestPoint<PointSource, PointTarget>::getUseReciprocalCorrespondences);\

#define BIND_SampleConsensusInitialAlignment(PointSource, PointTarget, FeatureT)                   \
  class_<pcl::SampleConsensusInitialAlignment<PointSource, PointTarget, FeatureT>,                 \
         base<pcl::Registration<PointSource, PointTarget>>>(                                       \
      "SampleConsensusInitialAlignment" #PointSource #PointTarget #FeatureT)                       \
      .constructor()                                                                               \
      .function("setInputSource", &pcl::SampleConsensusInitialAlignment<PointSource, PointTarget,  \
                                                                        FeatureT>::setInputSource) \
      .function("setInputTarget", &pcl::SampleConsensusInitialAlignment<PointSource, PointTarget,  \
                                                                        FeatureT>::setInputTarget) \
      .function("setSourceFeatures",                                                               \
                &pcl::SampleConsensusInitialAlignment<PointSource, PointTarget,                    \
                                                      FeatureT>::setSourceFeatures)                \
      .function("setTargetFeatures",                                                               \
                &pcl::SampleConsensusInitialAlignment<PointSource, PointTarget,                    \
                                                      FeatureT>::setTargetFeatures)                \
      .function("setMinSampleDistance",                                                            \
                &pcl::SampleConsensusInitialAlignment<PointSource, PointTarget,                    \
                                                      FeatureT>::setMinSampleDistance)             \
      .function("setNumberOfSamples",                                                              \
                &pcl::SampleConsensusInitialAlignment<PointSource, PointTarget,                    \
                                                      FeatureT>::setNumberOfSamples)               \
      .function("setCorrespondenceRandomness",                                                     \
                &pcl::SampleConsensusInitialAlignment<PointSource, PointTarget,                    \
                                                      FeatureT>::setCorrespondenceRandomness);

EMSCRIPTEN_BINDINGS(registration) {
  BIND_Registration(PointXYZ, PointXYZ);

  BIND_IterativeClosestPoint(PointXYZ, PointXYZ);

  BIND_SampleConsensusInitialAlignment(PointXYZ, PointXYZ, FPFHSignature33);


}

File: src/modules/registration/Registration.ts
import PCLBase from '@/modules/common/PCLBase';
import { PointCloud, wrapPointCloud } from '@/modules/common/PointCloud';
import { XYZPointTypes, PointXYZ } from '@/modules/common/point-types';
import { toXYZPointCloud } from '@/modules/common/utils';
import { Emscripten } from '@/types';

class Registration<T extends XYZPointTypes> extends PCLBase<T> {
  constructor(native: Emscripten.NativeAPI) {
    super(native);
  }

  /**
   * Only supports `PointXYZ` type,
   * if it is not `PointXYZ` type, it will use `toXYZPointCloud` method to convert to `PointXYZ`
   */
  public setInputSource(cloud: PointCloud<T>) {
    if (cloud._PT === PointXYZ) {
      this._native.setInputSource(cloud._native);
    } else {
      const cloudXYZ = new PointCloud<PointXYZ>();
      toXYZPointCloud(cloud, cloudXYZ);
      this._native.setInputSource(cloudXYZ._native);
      cloudXYZ.manager.delete();
    }
  }

  public getInputSource() {
    return wrapPointCloud<PointXYZ>(this._native.getInputSource());
  }

  /**
   * Only supports `PointXYZ` type,
   * if it is not `PointXYZ` type, it will use `toXYZPointCloud` method to convert to `PointXYZ`
   */
  public setInputTarget(cloud: PointCloud<T>) {
    if (cloud._PT === PointXYZ) {
      this._native.setInputTarget(cloud._native);
    } else {
      const cloudXYZ = new PointCloud<PointXYZ>();
      toXYZPointCloud(cloud, cloudXYZ);
      this._native.setInputTarget(cloudXYZ._native);
      cloudXYZ.manager.delete();
    }
  }

  public getInputTarget() {
    return wrapPointCloud<PointXYZ>(this._native.getInputTarget());
  }

  public getFitnessScore(maxRange = Number.MAX_VALUE): number {
    return this._native.getFitnessScore(maxRange);
  }

  public hasConverged(): boolean {
    return this._native.hasConverged();
  }

  public setMaximumIterations(maxIterations: number) {
    this._native.setMaximumIterations(maxIterations);
  }

  public getMaximumIterations(): number {
    return this._native.getMaximumIterations();
  }

  public setRANSACIterations(ransacIterations: number) {
    this._native.setRANSACIterations(ransacIterations);
  }

  public getRANSACIterations(): number {
    return this._native.getRANSACIterations();
  }

  public setRANSACOutlierRejectionThreshold(inlierThreshold: number) {
    this._native.setRANSACOutlierRejectionThreshold(inlierThreshold);
  }

  public getRANSACOutlierRejectionThreshold(): number {
    return this._native.getRANSACOutlierRejectionThreshold();
  }

  public setMaxCorrespondenceDistance(distanceThreshold: number) {
    this._native.setMaxCorrespondenceDistance(distanceThreshold);
  }

  public getMaxCorrespondenceDistance(): number {
    return this._native.getMaxCorrespondenceDistance();
  }

  public setTransformationEpsilon(epsilon: number) {
    this._native.setTransformationEpsilon(epsilon);
  }

  public getTransformationEpsilon(): number {
    return this._native.getTransformationEpsilon();
  }

  public setTransformationRotationEpsilon(epsilon: number) {
    this._native.setTransformationRotationEpsilon(epsilon);
  }

  public getTransformationRotationEpsilon(): number {
    return this._native.getTransformationRotationEpsilon();
  }

  public setEuclideanFitnessEpsilon(epsilon: number) {
    this._native.setEuclideanFitnessEpsilon(epsilon);
  }

  public getEuclideanFitnessEpsilon(): number {
    return this._native.getEuclideanFitnessEpsilon();
  }

  public initCompute(): boolean {
    return this._native.initCompute();
  }

  public initComputeReciprocal(): boolean {
    return this._native.initComputeReciprocal();
  }

  public align(output: PointCloud<T>,matrix: [] = []) {
    this._native.align(output._native,matrix);
  }

  public getFinalTransformation(){
    return this._native.getFinalTransformation();
  }
}

export default Registration;
